### DISCUSSION TITLE
We've just open-sourced a very competitive IO framework driven by io_uring !!!

### DISCUSSION THREAD

**Discussion by beef9999** [axboe/liburing#607] [Show and tell]
https://github.com/alibaba/PhotonLibOS

It's a production ready library while still being evolving. io_uring is working inside as the event engine, file io engine, and socket engine. Hopefully more people would join us...

Below are the performance data:

####  Disk IO

|        | IO Engine |  IO Type  | IO Size | IO Depth | DirectIO | QPS  | Throughput | CPU util |
|:------:|:---------:|:---------:|:-------:|:--------:|:--------:|:----:|:----------:|:--------:|
| Photon | io_uring  | Rand-read |   4KB   |   128    |   Yes    | 433K |   1.73GB   |   100%   |
| Photon |  libaio   | Rand-read |   4KB   |   128    |   Yes    | 346K |   1.38GB   |   100%   |
|  fio   |  libaio   | Rand-read |   4KB   |   128    |   Yes    | 279K |   1.11GB   |   100%   |

#### TCP echo server

Client Mode: Streaming

|                                                                       | Language / Coroutine Type | Buffer Size | Conn Num |  QPS  | Bandwidth | CPU util |
|:---------------------------------------------------------------------:|:-------------------------:|:-----------:|:--------:|:-----:|:---------:|:--------:|
|                                Photon                                 |      C++ / Stackful       |  512 Bytes  |    4     | 1604K |  6.12Gb   |   99%    |
|           [cocoyaxi](https://github.com/idealvin/cocoyaxi)            |      C++ / Stackful       |  512 Bytes  |    4     | 1545K |  5.89Gb   |   99%    |
|                      [tokio](https://tokio.rs/)                       |     Rust / Stackless      |  512 Bytes  |    4     | 1384K |  5.28Gb   |   98%    |
| [acl/lib_fiber](https://github.com/acl-dev/acl/tree/master/lib_fiber) |      C++ / Stackful       |  512 Bytes  |    4     | 1240K |  4.73Gb   |   94%    |
|                                  Go                                   |     Golang / Stackful     |  512 Bytes  |    4     | 1083K |  4.13Gb   |   100%   |
|              [libgo](https://github.com/yyzybb537/libgo)              |      C++ / Stackful       |  512 Bytes  |    4     | 770K  |  2.94Gb   |   99%    |
|             [boost::asio](https://think-async.com/Asio/)              |   C++ / Async-callback    |  512 Bytes  |    4     | 634K  |  2.42Gb   |   97%    |
|               [libco](https://github.com/Tencent/libco)               |      C++ / Stackful       |  512 Bytes  |    4     | 432K  |  1.65Gb   |   96%    |
|              [zab](https://github.com/Donald-Rupin/zab)               |     C++20 / Stackless     |  512 Bytes  |    4     | 412K  |  1.57Gb   |   99%    |
|             [asyncio](https://github.com/netcan/asyncio)              |     C++20 / Stackless     |  512 Bytes  |    4     | 163K  |  0.60Gb   |   98%    |

Client Mode: Ping-pong

|                                                                       | Language / Coroutine Type | Buffer Size | Conn Num | QPS  | Bandwidth | CPU util |
|:---------------------------------------------------------------------:|:-------------------------:|:-----------:|:--------:|:----:|:---------:|:--------:|
|                                Photon                                 |      C++ / Stackful       |  512 Bytes  |   1000   | 412K |  1.57Gb   |   100%   |
|             [boost::asio](https://think-async.com/Asio/)              |   C++ / Async-callback    |  512 Bytes  |   1000   | 393K |  1.49Gb   |   100%   |
|               [evpp](https://github.com/Qihoo360/evpp)                |   C++ / Async-callback    |  512 Bytes  |   1000   | 378K |  1.44Gb   |   100%   |
|                      [tokio](https://tokio.rs/)                       |     Rust / Stackless      |  512 Bytes  |   1000   | 365K |  1.39Gb   |   100%   |
|                                  Go                                   |     Golang / Stackful     |  512 Bytes  |   1000   | 331K |  1.26Gb   |   100%   |
| [acl/lib_fiber](https://github.com/acl-dev/acl/tree/master/lib_fiber) |      C++ / Stackful       |  512 Bytes  |   1000   | 327K |  1.25Gb   |   100%   |
|            [swoole](https://github.com/swoole/swoole-src)             |      PHP / Stackful       |  512 Bytes  |   1000   | 325K |  1.24Gb   |   99%    |
|              [zab](https://github.com/Donald-Rupin/zab)               |     C++20 / Stackless     |  512 Bytes  |   1000   | 317K |  1.21Gb   |   100%   |
|           [cocoyaxi](https://github.com/idealvin/cocoyaxi)            |      C++ / Stackful       |  512 Bytes  |   1000   | 279K |  1.06Gb   |   98%    |
|               [libco](https://github.com/Tencent/libco)               |      C++ / Stackful       |  512 Bytes  |   1000   | 260K |  0.99Gb   |   96%    |
|              [libgo](https://github.com/yyzybb537/libgo)              |      C++ / Stackful       |  512 Bytes  |   1000   | 258K |  0.98Gb   |   156%   |
|             [asyncio](https://github.com/netcan/asyncio)              |     C++20 / Stackless     |  512 Bytes  |   1000   | 142K |  0.54Gb   |   99%    |

#### HTTP

|        | File Size | QPS  | CPU util |
|:------:|:---------:|:----:|:--------:|
| Photon |    4KB    | 114K |   100%   |
| Nginx  |    4KB    | 97K  |   100%   |

**Comment by YoSTEALTH** [comment]
Congratulation on launching your project. HTTP 114K is that per thread?

**Reply by beef9999** [reply]
Yes, it is

**Reply by YoSTEALTH** [reply]
nice

**Comment by axboe** [comment]
Thanks for sharing this! Is this something Alibaba uses in production or is moving into production? Would be nice if you could talk about that aspect a little bit.

**Reply by beef9999** [reply]
Yes, this library is from the Storage Team of Alibaba Cloud, not personal work. Some of the cloud software are being using this lib for years. The async event engine used to be epoll, and we began to switch it to io_uring from last year. The io_uring socket still has performance issues, so we only use it to submit IO and do event processing.
 
Because our kernel team are responsible for maintaining the old 4.19 kernel and keep backporting 5.x io_uring features to 4.19, it makes us easier to persuade our customers (internal business teams) to start the io_uring trial. We hide the details under the coroutine library, so most of them won't be reluctant to cooperate with us doing the tests.

In a word, kernel is always the key point. If not given the fact that we have a production-ready 4.x kernel with io_uring features, I would suspect that it would take us a few more years to actually start looking at io_uring.

**Reply by axboe** [reply]
> The io_uring socket still has performance issues, so we only use it to submit IO and do event processing.

I'd love to work with  you on this, a lot of the recent work on io_uring has been networking focused. Both in terms of features, but also performance, and it's actively targeting some networking backends at Meta as well.

Can you expand on the socket performance issues you've been seeing?

**Reply by axboe** [reply]
A few examples of recent work that is actively improving the networking side for us:

1) The new ring provided buffers provide an efficient way to do a completion based socket model, something the previous provided buffers didn't quite do as they had a bit too much overhead for some situations.
2) The COOP_TASKRUN flag was specifically for some networking workloads that suffered from the forced reschedule that triggered whenever retry task_work got triggered.

The registered files improvements is another one that is generally useful like the above, but also specifically for sockets.

**Reply by beef9999** [reply]
For example, `blocking fd + iouring read/write` is slower then `non-blocking fd + iouring poll + psync read/write`. https://github.com/axboe/liburing/issues/536

But I heard people from the kernel teams said that the recent kernel had some improvements to reduce `io_uring_enter` cost. I didn't have a chance to try yet.

**Reply by axboe** [reply]
I wonder how close the strictly sync case is now. For most things that I test, we're doing better than epoll at this point, for some cases quite a bit more so. Recent changes for specifically io_uring_enter() is the registered ring fd, which also avoids the fget/fput for the ring fd itself (just like registered files do for IO files).

There's no reason why blocking fd + io_uring read/write should be slower than non-blocking fd + io_uring poll + psync read/write. Some things to keep in mind:

1) Ensure that the case doesn't hit io-wq. Any thread offload is a slower path in io_uring, obviously, and anything performant should not have any io-wq activity.
2) Don't use vectored reads/writes or send/receives if it can be avoided. Too many people default to using the vectored versions even for an iovec with 1 entry, and they are slower. Both for sync IO, but particularly with io_uring if we need to use internal poll, for example, to trigger a retry. We need to retain iovec state at that point, which is an alloc+copy of the vec. I've got some ideas for how to make the single vector readv/writev/recvmsg/sendmsg as good as the non-vectored case, but haven't implemented them yet.

**Reply by beef9999** [reply]
We don't have either of these two situations, maybe I can write a piece of POC code based on the newly open-sourced coroutine lib later.

**Reply by beef9999** [reply]
- The io-depth of a single connection is exactly 1 (can't be paralleled)
- When data is in memory, e.g. tmpfs or network buffer, the psync stack is more efficient than the io_uring stack

I think that's the major reason causing this situation (might be wrong)

**Reply by axboe** [reply]
It should be pretty close, if not identical, if done right. For example, a simple case of just reading sync from storage device. For this specific test, I get 102600KB/sec with dd and 102200KB/sec using io_uring. That's about a 0.4% difference, which is within respectable tolerances for this case imho. An async API, for fully sync stuff, is hard to make more efficient than the most basic of sync APIs.

For networking, this is usually non-blocking and readiness driven using epoll or something similar. That case is easier to gain efficiency for, as you need to do epoll, wait for readiness, get woken, pick buffer, issue receive. That's at least two system calls, and there's a bigger gap between readiness to receive than you would have with io_uring. For io_uring, it'd be issue recv, arms poll internally. Readiness triggers, receive is done, completion posted. The recv either has a buffer provided upfront, or it can use the ring mapped buffers to have it pick one at almost zero cost when we're ready to receive.

No backends handle a single connection with a single pending request, generally they handle many and you then additionally can get the benefit of handling multiple completions as a batch, and also issue them as a batch. This further cuts the per-socket overhead for receiving data.

There are further tricks you can do with the 5.19 codebase. By default, io_uring will attempt to always receive first for a recv request. If that fails, we rely on internal poll to trigger the retry. If you know (or have a strong assumption) that the socket will be empty when the receive is issued, you can ask io_uring to just go straight to poll. If that does say data is ready, then the receive happens inline anyway. This can be driven with looking at the previous receive for this socket, for example, which sets IORING_CQE_F_SOCK_NONEMPTY if there's data left in the socket after a receive.

IORING_CQE_F_SOCK_NONEMPTY also further helps to avoid the common situation where you epoll a socket for POLLIN, and when it's ready, you do a recv() or recvmsg() in a loop until you get -EAGAIN. This means that you always end up doing at least 1 system call extra to ensure you've read everything before you arm epoll for it again.

**Reply by beef9999** [reply]
I was testing psync and io_uring on a tmpfs file (/dev/shm), with one concurrency, and psync is 12.6GB/s and io_uring is 4.5GB/s.
So I think it's the memory backend that amplifies the difference.

**Reply by axboe** [reply]
Can you share the test code? That sounds like an absolutely massive difference, something is surely wrong.

**Reply by isilence** [reply]
> I was testing psync and io_uring on a tmpfs file (/dev/shm), with one concurrency, and psync is 12.6GB/s and io_uring is 4.5GB/s. So I think it's the memory backend that amplifies the difference.

Was it buffered writes? Performance of those is known to be bad and there is some WIP on that

https://lore.kernel.org/io-uring/20220616212221.2024518-1-shr@fb.com/

**Reply by axboe** [reply]
Hacked something up pretty quickly, here's what I see:

```
psync: 14110.36MB/sec
io_uring: 11715.93MB/sec
```

This is reading a 16GB file out of /dev/shm, psync is using pread() and io_uring is using io_uring_prep_read(). io_uring is using QD=1 and a ring of size 1 as well. All reads are 4k.

Adding a few small tricks (register the file and ring), we get to:

```
io_uring: 12087.33MB/sec
```

If we have both use 16KB reads instead, things get pretty close:

```
psync: 15738.48MB/sec
io_uring: 15155.57MB/sec
```

**Reply by axboe** [reply]
Test code
[sync-vs-iou.txt](https://github.com/axboe/liburing/files/8935532/sync-vs-iou.txt)
.

**Reply by axboe** [reply]
> Was it buffered writes? Performance of those is known to be bad and there is some WIP on that

Indeed, the writes is a totally different story, and the patchset from Stefan should boost that considerably. I was assuming we were just doing reads :-)

**Reply by axboe** [reply]
Given the initial perf differences described, it must be writes.

**Reply by beef9999** [reply]
No, I was testing reads, because writing a tmpfs file would somehow enable io-wq. Don't want it to get involved.

The tests were running in a cloud VM, with CentOS kernel 5.15.6

**Reply by axboe** [reply]
Then I'm a bit puzzled at the gap. We've made improvements since 5.15, but none that would explain a gap this big. Can you try my test app I attached in here? You'd need to remove the ring registration and the 1 << 8 from the queue init to run it on 5.15.

**Reply by beef9999** [reply]
After changing 1 << 8 to 0 and comment out io_uring_register_ring_fd, I got:

4KB:
psync: 7384.34MB/sec
io_uring: 6248.50MB/sec

16KB:
psync: 8590.87MB/sec
io_uring: 8000.58MB/sec

These data seem to be fine. So the only variable between your test code and mine is the `io_uring_register_files`

**Reply by beef9999** [reply]
After removing `io_uring_register_files `

4KB:
psync: 7381.09MB/sec
io_uring: 6123.07MB/sec

16KB:
psync: 8581.70MB/sec
io_uring: 8019.70MB/sec
